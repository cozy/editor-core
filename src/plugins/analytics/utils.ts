import { Step } from 'prosemirror-transform';
import { findParentNode } from 'prosemirror-utils';
import { CellSelection } from '@atlaskit/editor-tables/cell-selection';
import { EditorState, NodeSelection, Transaction } from 'prosemirror-state';
import { CreateUIAnalyticsEvent } from '@atlaskit/analytics-next';
import { InputRuleWithHandler } from '../../utils/input-rules';
import { GapCursorSelection, Side } from '../selection/gap-cursor/selection';
import { AnalyticsStep } from '@atlaskit/adf-schema/steps';
import { editorAnalyticsChannel } from './consts';
import {
  AnalyticsEventPayloadWithChannel,
  AnalyticsEventPayload,
  ACTION,
  ACTION_SUBJECT,
  TABLE_ACTION,
} from './types';
import { SELECTION_TYPE, SELECTION_POSITION } from './types/utils';
import { analyticsPluginKey } from './plugin-key';
import { HigherOrderCommand } from '../../types/command';
import { fireAnalyticsEvent } from './fire-analytics-event';

function getCreateUIAnalyticsEvent(
  editorState: EditorState,
): CreateUIAnalyticsEvent | null | undefined {
  const pluginState = analyticsPluginKey.getState(editorState);
  return pluginState
    ? (pluginState.createAnalyticsEvent as
        | CreateUIAnalyticsEvent
        | null
        | undefined)
    : undefined;
}

export function getStateContext(
  state: EditorState,
  payload: AnalyticsEventPayload,
): AnalyticsEventPayload {
  if (!payload.attributes) {
    return payload;
  }

  const { type, position } = getSelectionType(state);
  payload.attributes.selectionType = type;
  if (position) {
    payload.attributes.selectionPosition = position;
  }
  const insertLocation = findInsertLocation(state);

  if (
    payload.action === ACTION.INSERTED &&
    payload.actionSubject === ACTION_SUBJECT.DOCUMENT &&
    payload.attributes
  ) {
    payload.attributes.insertLocation = insertLocation;
  } else {
    payload.attributes.nodeLocation = insertLocation;
  }

  return payload;
}

export function getSelectionType(
  state: EditorState,
): { type: SELECTION_TYPE; position?: SELECTION_POSITION } {
  const { selection } = state;
  let type: SELECTION_TYPE;
  let position: SELECTION_POSITION | undefined;

  if (selection instanceof GapCursorSelection) {
    type = SELECTION_TYPE.GAP_CURSOR;
    position =
      selection.side === Side.LEFT
        ? SELECTION_POSITION.LEFT
        : SELECTION_POSITION.RIGHT;
  } else if (selection instanceof CellSelection) {
    type = SELECTION_TYPE.CELL;
  } else if (selection instanceof NodeSelection) {
    type = SELECTION_TYPE.NODE;
  } else if (selection.from !== selection.to) {
    type = SELECTION_TYPE.RANGED;
  } else {
    type = SELECTION_TYPE.CURSOR;
    const { from, $from } = selection;
    if (from === $from.start()) {
      position = SELECTION_POSITION.START;
    } else if (from === $from.end()) {
      position = SELECTION_POSITION.END;
    } else {
      position = SELECTION_POSITION.MIDDLE;
    }
  }

  return {
    type,
    position,
  };
}

export function findInsertLocation(state: EditorState): string {
  const { selection } = state;
  if (selection instanceof NodeSelection) {
    return selection.node.type.name;
  }

  if (selection instanceof CellSelection) {
    return state.schema.nodes.table.name;
  }

  // Text selection
  const parentNodeInfo = findParentNode(
    node => node.type !== state.schema.nodes.paragraph,
  )(state.selection);

  return parentNodeInfo ? parentNodeInfo.node.type.name : state.doc.type.name;
}

const actionsToIgnore: (ACTION | TABLE_ACTION)[] = [
  ACTION.INVOKED,
  ACTION.OPENED,
];

export function addAnalytics(
  state: EditorState,
  tr: Transaction,
  payload: AnalyticsEventPayload,
  channel: string = editorAnalyticsChannel,
): Transaction {
  const createAnalyticsEvent = getCreateUIAnalyticsEvent(state);
  payload = getStateContext(state, payload);

  if (createAnalyticsEvent) {
    const { storedMarks } = tr;
    tr.step(
      new AnalyticsStep(
        (payload: AnalyticsEventPayload, channel: string) => {
          fireAnalyticsEvent(createAnalyticsEvent)({ payload, channel });
        },
        [
          {
            payload,
            channel,
          },
        ],
        actionsToIgnore,
        tr.selection.$from.pos, // We need to create the step based on a position, this prevent split history for relative changes.
      ),
    );
    // When you add a new step all the storedMarks are removed it
    if (storedMarks) {
      tr.setStoredMarks(storedMarks);
    }
  }

  return tr;
}

export type AnalyticsEventPayloadCallback = (
  state: EditorState,
) => AnalyticsEventPayload | undefined;

export function withAnalytics(
  payload: AnalyticsEventPayload | AnalyticsEventPayloadCallback,
  channel?: string,
): HigherOrderCommand {
  return command => (state, dispatch, view) =>
    command(
      state,
      tr => {
        if (dispatch) {
          if (payload instanceof Function) {
            const dynamicPayload = payload(state);
            if (dynamicPayload) {
              dispatch(addAnalytics(state, tr, dynamicPayload, channel));
            }
          } else {
            dispatch(addAnalytics(state, tr, payload, channel));
          }
        }
      },
      view,
    );
}

export function ruleWithAnalytics(
  getPayload: (
    state: EditorState,
    match: string[],
    start: number,
    end: number,
  ) => AnalyticsEventPayload,
) {
  return (rule: InputRuleWithHandler) => {
    // Monkey patching handler to add analytics
    const handler = rule.handler;

    rule.handler = (
      state: EditorState,
      match,
      start,
      end,
    ): Transaction<any> | null => {
      let tr = handler(state, match, start, end);

      if (tr) {
        const payload = getPayload(state, match, start, end);
        tr = addAnalytics(state, tr, payload);
      }
      return tr;
    };
    return rule;
  };
}

export function getAnalyticsEventsFromTransaction(
  tr: Transaction,
): AnalyticsEventPayloadWithChannel[] {
  return (tr.steps as Step[])
    .filter<AnalyticsStep<AnalyticsEventPayload>>(
      (step: Step): step is AnalyticsStep<AnalyticsEventPayload> =>
        step instanceof AnalyticsStep,
    )
    .reduce<AnalyticsEventPayloadWithChannel[]>(
      (acc, step) => [...acc, ...step.analyticsEvents],
      [],
    );
}
