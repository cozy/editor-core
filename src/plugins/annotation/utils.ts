import type {
  ResolvedPos,
  Mark,
  Node,
  Slice,
  Schema,
} from '@atlaskit/editor-prosemirror/model';
import type {
  EditorState,
  Selection,
} from '@atlaskit/editor-prosemirror/state';
import {
  PluginKey,
  TextSelection,
  AllSelection,
} from '@atlaskit/editor-prosemirror/state';
import { Decoration } from '@atlaskit/editor-prosemirror/view';
import { AnnotationSharedClassNames } from '@atlaskit/editor-common/styles';
import {
  canApplyAnnotationOnRange,
  getAnnotationIdsFromRange,
  isText,
  isParagraph,
  hasAnnotationMark,
  containsAnyAnnotations,
} from '@atlaskit/editor-common/utils';
import type { AnnotationMarkAttributes } from '@atlaskit/adf-schema';
import { AnnotationTypes } from '@atlaskit/adf-schema';
import type { AnnotationInfo } from './types';
import { AnnotationSelectionType } from './types';
import type { InlineCommentPluginState } from './pm-plugins/types';
import {
  ACTION_SUBJECT,
  ACTION_SUBJECT_ID,
  EVENT_TYPE,
  ACTION,
} from '@atlaskit/editor-common/analytics';
import type { AnalyticsEventPayloadCallback } from '@atlaskit/editor-common/analytics';
import type {
  INPUT_METHOD,
  AnnotationAEPAttributes,
  AnalyticsEventPayload,
} from '@atlaskit/editor-common/analytics';

export { hasAnnotationMark, containsAnyAnnotations };
function sum<T>(arr: Array<T>, f: (val: T) => number) {
  return arr.reduce((val, x) => val + f(x), 0);
}
/**
 * Finds the marks in the nodes to the left and right.
 * @param $pos Position to center search around
 */
export const surroundingMarks = ($pos: ResolvedPos) => {
  const { nodeBefore, nodeAfter } = $pos;
  const markNodeBefore =
    nodeBefore &&
    $pos.doc.nodeAt(Math.max(0, $pos.pos - nodeBefore.nodeSize - 1));
  const markNodeAfter =
    nodeAfter && $pos.doc.nodeAt($pos.pos + nodeAfter.nodeSize);

  return [
    (markNodeBefore && markNodeBefore.marks) || [],
    (markNodeAfter && markNodeAfter.marks) || [],
  ];
};

/**
 * Finds annotation marks, and returns their IDs.
 * @param marks Array of marks to search in
 */
const filterAnnotationIds = (marks: readonly Mark[]): Array<string> => {
  if (!marks.length) {
    return [];
  }

  const { annotation } = marks[0].type.schema.marks;
  return marks
    .filter((mark) => mark.type === annotation)
    .map((mark) => mark.attrs.id);
};

/**
 * Re-orders the annotation array based on the order in the document.
 *
 * This places the marks that do not appear in the surrounding nodes
 * higher in the list. That is, the inner-most one appears first.
 *
 * Undo, for example, can re-order annotation marks in the document.
 * @param annotations annotation metadata
 * @param $from location to look around (usually the selection)
 */
const reorderAnnotations = (
  annotations: Array<AnnotationInfo>,
  $from: ResolvedPos,
) => {
  const idSet = surroundingMarks($from).map(filterAnnotationIds);

  annotations.sort(
    (a, b) =>
      sum(idSet, (ids) => ids.indexOf(a.id)) -
      sum(idSet, (ids) => ids.indexOf(b.id)),
  );
};

export const getAllAnnotations = (doc: Node): string[] => {
  const allAnnotationIds: Set<string> = new Set();

  doc.descendants((node) => {
    node.marks
      .filter((mark) => mark.type.name === 'annotation')
      // filter out annotations with invalid attributes as they cause errors when interacting with them
      .filter(validateAnnotationMark)
      .forEach((m) => allAnnotationIds.add(m.attrs.id));
    return true;
  });

  return Array.from(allAnnotationIds);
};

/*
 * verifies if annotation mark contains valid attributes
 */
const validateAnnotationMark = (annotationMark: Mark): boolean => {
  const { id, annotationType } =
    annotationMark.attrs as AnnotationMarkAttributes;
  return validateAnnotationId(id) && validateAnnotationType(annotationType);

  function validateAnnotationId(id: string): boolean {
    if (!id || typeof id !== 'string') {
      return false;
    }
    const invalidIds = ['null', 'undefined'];
    return !invalidIds.includes(id.toLowerCase());
  }

  function validateAnnotationType(type: AnnotationTypes): boolean {
    if (!type || typeof type !== 'string') {
      return false;
    }
    const allowedTypes = Object.values(AnnotationTypes);
    return allowedTypes.includes(type);
  }
};

/*
 * add decoration for the comment selection in draft state
 * (when creating new comment)
 */
export const addDraftDecoration = (start: number, end: number) => {
  return Decoration.inline(start, end, {
    class: `${AnnotationSharedClassNames.draft}`,
  });
};

export const getAnnotationViewKey = (annotations: AnnotationInfo[]): string => {
  const keys = annotations.map((mark) => mark.id).join('_');
  return `view-annotation-wrapper_${keys}`;
};

export const findAnnotationsInSelection = (
  selection: Selection,
  doc: Node,
): AnnotationInfo[] => {
  const { empty, $anchor, anchor } = selection;
  // Only detect annotations on caret selection
  if (!empty || !doc) {
    return [];
  }

  const node = doc.nodeAt(anchor);
  if (!node && !$anchor.nodeBefore) {
    return [];
  }

  const annotationMark = doc.type.schema.marks.annotation;
  const nodeBefore = $anchor.nodeBefore;
  const anchorAnnotationMarks = (node && node.marks) || [];

  let marks: readonly Mark[] = [];
  if (annotationMark.isInSet(anchorAnnotationMarks)) {
    marks = anchorAnnotationMarks;
  } else if (nodeBefore && annotationMark.isInSet(nodeBefore.marks)) {
    marks = nodeBefore.marks;
  }

  const annotations = marks
    .filter((mark) => mark.type.name === 'annotation')
    .map((mark) => ({
      id: mark.attrs.id,
      type: mark.attrs.annotationType,
    }));

  reorderAnnotations(annotations, $anchor);
  return annotations;
};

/**
 * get selection from position to apply new comment for
 * @return bookmarked positions if they exists, otherwise current selection positions
 */
export function getSelectionPositions(
  editorState: EditorState,
  inlineCommentState?: InlineCommentPluginState | null | undefined,
): Selection {
  const { bookmark } = inlineCommentState || {};
  // get positions via saved bookmark if it is available
  // this is to make comments box positioned relative to temporary highlight rather then current selection
  if (bookmark) {
    return bookmark.resolve(editorState.doc);
  }
  return editorState.selection;
}

export const inlineCommentPluginKey = new PluginKey<InlineCommentPluginState>(
  'inlineCommentPluginKey',
);

export const getPluginState = (state: EditorState) => {
  return inlineCommentPluginKey.getState(state);
};

/**
 * get number of unique annotations within current selection
 */
const getAnnotationsInSelectionCount = (state: EditorState): number => {
  const { from, to } = state.selection;
  const annotations = getAnnotationIdsFromRange(
    { from, to },
    state.doc,
    state.schema,
  );
  return annotations.length;
};

/**
 * get payload for the open/close analytics event
 */
export const getDraftCommandAnalyticsPayload = (
  drafting: boolean,
  inputMethod: INPUT_METHOD.TOOLBAR | INPUT_METHOD.SHORTCUT,
) => {
  const payload: AnalyticsEventPayloadCallback = (
    state: EditorState,
  ): AnalyticsEventPayload => {
    let attributes: AnnotationAEPAttributes = {};

    if (drafting) {
      attributes = {
        inputMethod,
        overlap: getAnnotationsInSelectionCount(state),
      };
    }

    return {
      action: drafting ? ACTION.OPENED : ACTION.CLOSED,
      actionSubject: ACTION_SUBJECT.ANNOTATION,
      actionSubjectId: ACTION_SUBJECT_ID.INLINE_COMMENT,
      eventType: EVENT_TYPE.TRACK,
      attributes,
    };
  };
  return payload;
};

export const isSelectionValid = (
  state: EditorState,
): AnnotationSelectionType => {
  const { selection } = state;
  const { disallowOnWhitespace } = getPluginState(state) || {};

  if (
    selection.empty ||
    !(selection instanceof TextSelection || selection instanceof AllSelection)
  ) {
    return AnnotationSelectionType.INVALID;
  }

  const containsInvalidNodes = hasInvalidNodes(state);

  // A selection that only covers 1 pos, and is an invalid node
  // e.g. a text selection over a mention
  if (containsInvalidNodes && selection.to - selection.from === 1) {
    return AnnotationSelectionType.INVALID;
  }

  if (containsInvalidNodes) {
    return AnnotationSelectionType.DISABLED;
  }

  if (
    disallowOnWhitespace &&
    hasInvalidWhitespaceNode(selection, state.schema)
  ) {
    return AnnotationSelectionType.INVALID;
  }

  if (isEmptyTextSelection(selection, state.schema)) {
    return AnnotationSelectionType.INVALID;
  }

  return AnnotationSelectionType.VALID;
};

export const hasInvalidNodes = (state: EditorState): boolean => {
  const { selection, doc, schema } = state;
  return !canApplyAnnotationOnRange(
    {
      from: selection.from,
      to: selection.to,
    },
    doc,
    schema,
  );
};

/**
 * Checks if selection contains only empty text
 * e.g. when you select across multiple empty paragraphs
 */
function isEmptyTextSelection(
  selection: TextSelection | AllSelection,
  schema: any,
) {
  const { text, paragraph } = schema.nodes;
  let hasContent = false;
  selection.content().content.descendants((node) => {
    // for empty paragraph - consider empty (nothing to comment on)
    if (node.type === paragraph && !node.content.size) {
      return false;
    }
    // for not a text or nonempty text - consider nonempty (can comment if the node is supported for annotations)
    if (node.type !== text || !node.textContent) {
      hasContent = true;
    }
    return !hasContent;
  });
  return !hasContent;
}

/**
 * Checks if any of the nodes in a given selection are completely whitespace
 * This is to conform to Confluence annotation specifications
 */
export function hasInvalidWhitespaceNode(
  selection: TextSelection | AllSelection,
  schema: Schema,
) {
  let foundInvalidWhitespace = false;

  const content = selection.content().content;

  content.descendants((node) => {
    if (isText(node, schema)) {
      return false;
    }
    if (node.textContent.trim() === '') {
      // Trailing new lines do not result in the annotation spanning into
      // the trailing new line so can be ignored when looking for invalid
      // whitespace nodes.
      const nodeIsTrailingNewLine =
        // it is the final node
        node.eq(content.lastChild!) &&
        // and there are multiple nodes
        !node.eq(content.firstChild!) &&
        // and it is a paragraph node
        isParagraph(node, schema);

      if (!nodeIsTrailingNewLine) {
        foundInvalidWhitespace = true;
      }
    }

    return !foundInvalidWhitespace;
  });
  return foundInvalidWhitespace;
}

/*
 * verifies that the annotation exists by the given id
 */
export function annotationExists(
  annotationId: string,
  state: EditorState,
): boolean {
  const commentsPluginState = getPluginState(state);
  return (
    !!commentsPluginState?.annotations &&
    Object.keys(commentsPluginState.annotations).includes(annotationId)
  );
}

/*
 * remove annotations that dont exsist in plugin state from slice
 */
export function stripNonExistingAnnotations(slice: Slice, state: EditorState) {
  if (!slice.content.size) {
    return false;
  }
  slice.content.forEach((node) => {
    stripNonExistingAnnotationsFromNode(node, state);
    node.content.descendants((node) => {
      stripNonExistingAnnotationsFromNode(node, state);
      return true;
    });
  });
}

/*
 * remove annotations that dont exsist in plugin state
 * from node
 */
function stripNonExistingAnnotationsFromNode(node: Node, state: EditorState) {
  if (hasAnnotationMark(node, state)) {
    (node.marks as Mark[]) = node.marks.filter((mark) => {
      if (mark.type.name === 'annotation') {
        return annotationExists(mark.attrs.id, state);
      }
      return true;
    });
  }
  return node;
}
