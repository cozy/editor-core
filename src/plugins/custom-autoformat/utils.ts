import type {
  EditorState,
  Transaction,
} from '@atlaskit/editor-prosemirror/state';
import { PluginKey } from '@atlaskit/editor-prosemirror/state';

import type { CustomAutoformatState, CustomAutoformatAction } from './types';

export const pluginKey = new PluginKey('customAutoformatPlugin');

export const getPluginState = (editorState: EditorState) =>
  pluginKey.getState(editorState) as CustomAutoformatState | undefined;

export const autoformatAction = (
  tr: Transaction,
  action: CustomAutoformatAction,
): Transaction => {
  return tr.setMeta(pluginKey, action);
};
