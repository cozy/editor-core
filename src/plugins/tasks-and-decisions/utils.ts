import type {
  Node as PMNode,
  ResolvedPos,
} from '@atlaskit/editor-prosemirror/model';
import {
  NodeSelection,
  Selection,
  TextSelection,
} from '@atlaskit/editor-prosemirror/state';
import { transformSliceToDecisionList } from '@atlaskit/editor-common/transforms';

export { transformSliceToDecisionList };

function isTaskListNode(node: PMNode | null | undefined) {
  return Boolean(node && node.type && 'taskList' === node.type.name);
}

const resolvePositionToStartOfTaskItem = ($pos: ResolvedPos): ResolvedPos => {
  const fromRange = $pos.blockRange($pos, isTaskListNode);
  const fromPosition =
    fromRange && $pos.textOffset === 0 && fromRange.end - 1 === $pos.pos
      ? Selection.near($pos.doc.resolve(fromRange.end + 1), 1).$from
      : $pos;
  return fromPosition;
};

const resolvePositionToEndOfTaskItem = ($pos: ResolvedPos): ResolvedPos => {
  const toRange = $pos.blockRange($pos, isTaskListNode);
  const toPosition =
    toRange && $pos.textOffset === 0 && toRange.start + 1 === $pos.pos
      ? Selection.near($pos.doc.resolve(toRange.start - 1), -1).$to
      : $pos;

  return toPosition;
};

export const normalizeTaskItemsSelection = (
  selection: Selection,
): Selection => {
  if (selection.empty) {
    return selection;
  }

  const { $from, $to } = selection;

  if (selection instanceof NodeSelection) {
    const head = resolvePositionToStartOfTaskItem($from);
    return new TextSelection(head, head);
  }

  const head = resolvePositionToStartOfTaskItem($from);
  const anchor = resolvePositionToEndOfTaskItem($to);

  return new TextSelection(anchor, head);
};
