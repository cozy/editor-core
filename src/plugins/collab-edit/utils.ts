import type { EditorState } from '@atlaskit/editor-prosemirror/state';
import { Selection, TextSelection } from '@atlaskit/editor-prosemirror/state';
import type {
  EditorView,
  DecorationSet,
} from '@atlaskit/editor-prosemirror/view';
import { Decoration } from '@atlaskit/editor-prosemirror/view';
import type { Node as PMNode } from '@atlaskit/editor-prosemirror/model';
import * as themeColors from '@atlaskit/theme/colors';

import { hexToRgba } from '@atlaskit/adf-schema';
import { ZERO_WIDTH_JOINER } from '@atlaskit/editor-common/utils';
import type {
  AnalyticsEventPayload,
  EditorAnalyticsAPI,
} from '@atlaskit/editor-common/analytics';
import {
  EVENT_TYPE,
  ACTION,
  ACTION_SUBJECT,
} from '@atlaskit/editor-common/analytics';

import type {
  CollabParticipant,
  CollabEditOptions,
} from '@atlaskit/editor-common/collab';

export interface Color {
  solid: string;
  selection: string;
}

// TODO: https://product-fabric.atlassian.net/browse/DSP-7269
/* eslint-disable @atlaskit/design-system/ensure-design-token-usage */
export const colors: Color[] = [
  themeColors.R100,
  themeColors.R300,
  themeColors.R500,
  themeColors.Y100,
  themeColors.Y300,
  themeColors.Y500,
  themeColors.G100,
  themeColors.G300,
  themeColors.G500,
  themeColors.T100,
  themeColors.T300,
  themeColors.T500,
  themeColors.B100,
  themeColors.B300,
  themeColors.B500,
  themeColors.P100,
  themeColors.P300,
  themeColors.P500,
  themeColors.N70,
  themeColors.N200,
  themeColors.N800,
].map((solid) => ({
  solid,
  selection: hexToRgba(solid, 0.2)!,
}));
/* eslint-enable @atlaskit/design-system/ensure-design-token-usage */

export const getAvatarColor = (str: string) => {
  let hash = 0;

  for (let i = 0; i < str.length; i++) {
    /* eslint-disable no-bitwise */
    hash = (hash << 5) - hash + str.charCodeAt(i);
    hash = hash & hash;
    /* eslint-enable no-bitwise */
  }

  const index = Math.abs(hash) % colors.length;

  return { index, color: colors[index] };
};

export const findPointers = (
  id: string,
  decorations: DecorationSet,
): Decoration[] =>
  decorations
    .find()
    .reduce(
      (arr, deco: any) =>
        deco.spec.pointer.sessionId === id ? arr.concat(deco) : arr,
      [],
    );

function style(options: { color: string }) {
  const color = (options && options.color) || 'black';
  return `border-left: 1px solid ${color}; border-right: 1px solid ${color}; margin-right: -2px;`;
}

export const createTelepointers = (
  from: number,
  to: number,
  sessionId: string,
  isSelection: boolean,
  initial: string,
) => {
  let decorations: Decoration[] = [];
  const avatarColor = getAvatarColor(sessionId);
  const color = avatarColor.index.toString();
  if (isSelection) {
    const className = `telepointer color-${color} telepointer-selection`;
    decorations.push(
      (Decoration as any).inline(
        from,
        to,
        { class: className, 'data-initial': initial },
        { pointer: { sessionId } },
      ),
    );
  }

  const spaceJoinerBefore = document.createElement('span');
  spaceJoinerBefore.textContent = ZERO_WIDTH_JOINER;
  const spaceJoinerAfter = document.createElement('span');
  spaceJoinerAfter.textContent = ZERO_WIDTH_JOINER;

  const cursor = document.createElement('span');
  cursor.textContent = ZERO_WIDTH_JOINER;
  cursor.className = `telepointer color-${color} telepointer-selection-badge`;
  cursor.style.cssText = `${style({ color: avatarColor.color.solid })};`;
  cursor.setAttribute('data-initial', initial);
  return decorations
    .concat(
      (Decoration as any).widget(to, spaceJoinerAfter, {
        pointer: { sessionId },
        key: `telepointer-${sessionId}-zero`,
      }),
    )
    .concat(
      (Decoration as any).widget(to, cursor, {
        pointer: { sessionId },
        key: `telepointer-${sessionId}`,
      }),
    )
    .concat(
      (Decoration as any).widget(to, spaceJoinerBefore, {
        pointer: { sessionId },
        key: `telepointer-${sessionId}-zero`,
      }),
    );
};

export const replaceDocument = (
  doc: any,
  state: EditorState,
  version?: number,
  options?: CollabEditOptions,
  reserveCursor?: boolean,
) => {
  const { schema, tr } = state;

  let content: Array<PMNode> = (doc.content || []).map((child: any) =>
    schema.nodeFromJSON(child),
  );
  let hasContent = !!content.length;

  if (hasContent) {
    tr.setMeta('addToHistory', false);
    tr.replaceWith(0, state.doc.nodeSize - 2, content!);
    const selection = state.selection;
    if (reserveCursor) {
      // If the cursor is still in the range of the new document,
      // keep where it was.
      if (selection.to < tr.doc.content.size - 2) {
        const $from = tr.doc.resolve(selection.from);
        const $to = tr.doc.resolve(selection.to);
        const newselection = new TextSelection($from, $to);
        tr.setSelection(newselection);
      }
    } else {
      tr.setSelection(Selection.atStart(tr.doc));
    }
    tr.setMeta('replaceDocument', true);

    if (typeof version !== undefined && options && options.useNativePlugin) {
      const collabState = { version, unconfirmed: [] };
      tr.setMeta('collab$', collabState);
    }
  }

  return tr;
};

export const scrollToCollabCursor = (
  editorView: EditorView,
  participants: CollabParticipant[],
  sessionId: string | undefined,
  // analytics: AnalyticsEvent | undefined,
  index: number,
  editorAnalyticsAPI: EditorAnalyticsAPI | undefined,
) => {
  const selectedUser = participants[index];
  if (
    selectedUser &&
    selectedUser.cursorPos !== undefined &&
    selectedUser.sessionId !== sessionId
  ) {
    const { state } = editorView;
    let tr = state.tr;
    const analyticsPayload: AnalyticsEventPayload = {
      action: ACTION.MATCHED,
      actionSubject: ACTION_SUBJECT.SELECTION,
      eventType: EVENT_TYPE.TRACK,
    };
    tr.setSelection(Selection.near(tr.doc.resolve(selectedUser.cursorPos)));
    editorAnalyticsAPI?.attachAnalyticsEvent(analyticsPayload)(tr);
    tr.scrollIntoView();
    editorView.dispatch(tr);
    if (!editorView.hasFocus()) {
      editorView.focus();
    }
  }
};

export const getPositionOfTelepointer = (
  sessionId: string,
  decorationSet: DecorationSet,
): undefined | number => {
  let scrollPosition;
  decorationSet.find().forEach((deco: any) => {
    if (deco.type.spec.pointer.sessionId === sessionId) {
      scrollPosition = deco.from;
    }
  });
  return scrollPosition;
};
