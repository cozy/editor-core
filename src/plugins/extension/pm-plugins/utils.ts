import type { EditorView } from '@atlaskit/editor-prosemirror/view';
import type { ExtensionProvider } from '@atlaskit/editor-common/extensions';
import { getExtensionModuleNode } from '@atlaskit/editor-common/extensions';
import { updateState } from '../commands';
import { getSelectedExtension } from '../utils';

const maybeGetUpdateMethodFromExtensionProvider = async (
  view: EditorView,
  extensionProvider: ExtensionProvider,
) => {
  const nodeWithPos = getSelectedExtension(view.state, true);
  if (!nodeWithPos) {
    throw new Error('There is no selection');
  }

  const { extensionType, extensionKey } = nodeWithPos.node.attrs;

  const extensionModuleNode = await getExtensionModuleNode(
    extensionProvider,
    extensionType,
    extensionKey,
  );

  const newNodeWithPos = getSelectedExtension(view.state, true);

  if (
    newNodeWithPos &&
    newNodeWithPos.node.attrs.extensionType === extensionType &&
    newNodeWithPos.node.attrs.extensionKey === extensionKey &&
    newNodeWithPos.pos === nodeWithPos.pos &&
    extensionModuleNode
  ) {
    return extensionModuleNode.update;
  }
};

export const updateEditButton = async (
  view: EditorView,
  extensionProvider: ExtensionProvider,
) => {
  try {
    const updateMethod = await maybeGetUpdateMethodFromExtensionProvider(
      view,
      extensionProvider,
    );

    updateState({
      showEditButton: !!updateMethod,
      updateExtension:
        (updateMethod && Promise.resolve(updateMethod)) || undefined,
    })(view.state, view.dispatch);

    return updateMethod;
  } catch {
    // this exception is not important for this case, fail silently
  }
};
