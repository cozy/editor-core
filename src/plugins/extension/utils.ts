import { EditorState } from 'prosemirror-state';
import {
  findParentNodeOfType,
  findSelectedNodeOfType,
  findDomRefAtPos,
  NodeWithPos,
  DomAtPos,
} from 'prosemirror-utils';
import { closestElement } from '../../utils/dom';

export const getSelectedExtension = (
  state: EditorState,
  searchParent: boolean = false,
) => {
  const { inlineExtension, extension, bodiedExtension } = state.schema.nodes;
  const nodeTypes = [extension, bodiedExtension, inlineExtension];
  return (
    findSelectedNodeOfType(nodeTypes)(state.selection) ||
    (searchParent && findParentNodeOfType(nodeTypes)(state.selection)) ||
    undefined
  );
};

export const getSelectedNonContentExtension = ({
  schema,
  selection,
}: EditorState): NodeWithPos | undefined => {
  const { inlineExtension, extension } = schema.nodes;
  return findSelectedNodeOfType([inlineExtension, extension])(selection);
};

export const getSelectedDomElement = (
  domAtPos: DomAtPos,
  selectedExtensionNode?: NodeWithPos,
  isContentExtension: boolean = false,
) => {
  const selectedExtensionDomNode =
    selectedExtensionNode &&
    (findDomRefAtPos(selectedExtensionNode.pos, domAtPos) as HTMLElement);

  // Non-content extension can be nested in bodied-extension, the following check is necessary for that case
  return selectedExtensionNode && selectedExtensionDomNode!.querySelector
    ? isContentExtension
      ? selectedExtensionDomNode!.querySelector<HTMLElement>(
          '.extension-container',
        ) || selectedExtensionDomNode
      : closestElement(selectedExtensionDomNode!, '.extension-container') ||
        selectedExtensionDomNode!.querySelector<HTMLElement>(
          '.extension-container',
        ) ||
        selectedExtensionDomNode
    : undefined;
};
