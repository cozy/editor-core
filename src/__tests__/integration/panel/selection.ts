import { BrowserTestCase } from '@atlaskit/webdriver-runner/runner';

import {
  goToEditorTestingExample,
  mountEditor,
} from '../../__helpers/testing-example-helpers';
import { panelSelectors } from '../../__helpers/page-objects/_panel';
import {
  editable,
  expectToMatchDocument,
  fullpage,
  getProsemirrorSelection,
} from '../_helpers';

import panelAdf from './__fixtures__/basic-panel-adf.json';
import { calcUserDragAndDropFromMidPoint } from '../../__helpers/utils';

BrowserTestCase(
  'selection.ts: Writing inside the panel, selecting panel and typing should drop text in panel',
  {},
  async (client: any, testName: string) => {
    const page = await goToEditorTestingExample(client);
    await mountEditor(page, {
      appearance: fullpage.appearance,
      allowPanel: true,
      defaultValue: panelAdf,
    });

    // Select the left margin of the panel, selecting the node
    await page.click(panelSelectors.icon);
    await page.type(editable, 'this text should be in doc root');

    await expectToMatchDocument(page, testName);
  },
);

BrowserTestCase(
  "doesn't select panel node if click and drag before releasing mouse",
  {},
  async (client: any) => {
    const page = await goToEditorTestingExample(client);
    await mountEditor(page, {
      appearance: fullpage.appearance,
      allowPanel: true,
      defaultValue: panelAdf,
    });

    // click and drag from centre of panel out to padding
    const boundingRect = await page.getBoundingRect(panelSelectors.panel);
    await page.simulateUserDragAndDrop(
      ...calcUserDragAndDropFromMidPoint(boundingRect, 5),
    );

    const prosemirrorSelection = await getProsemirrorSelection(page);
    if (!prosemirrorSelection) {
      throw new Error('Unable to get Prosemirror selection');
    }
    expect(prosemirrorSelection.type).not.toBe('node');
  },
);
