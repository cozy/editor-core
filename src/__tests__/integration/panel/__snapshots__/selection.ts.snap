// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`selection.ts: Writing inside the panel, selecting panel and typing should drop text in panel 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "text": "this text should be in doc root",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
  "version": 1,
}
`;
